src/common/datomish/transact.cljc

```diff
@@ -552,12 +580,99 @@
     (:db-after (<? (<with db tx-data)))))
 
 (defn <transact!
-  [conn tx-data]
+  "Submits a transaction to the database for writing.
+
+  Returns a pair-chan resolving to `[result error]`."
+  ([conn tx-data]
+   (<transact! conn tx-data (a/chan 1) true))
+  ([conn tx-data result close?]
+   {:pre [(conn? conn)]}
+   ;; Any race to put! is a real race between callers of <transact!.  We can't just park on put!,
+   ;; because the parked putter that is woken is non-deterministic.
+   (let [closed? (not (a/put! (:transact-chan conn) [:sentinel-transact tx-data result close?]))]
+     (go-pair
+       ;; We want to return a pair-chan, no matter what kind of channel result is.
+       (if closed?
+         (raise "Connection is closed" {:error :transact/connection-closed})
+         (<? result))))))
+
+(defn- start-transactor [conn]
+  (let [token-chan (a/chan 1)]
+    (go
+      (>! token-chan (gensym "transactor-token"))
+      (loop []
+        (when-let [token (<! token-chan)]
+          (when-let [[sentinel tx-data result close?] (<! (:transact-chan conn))]
```

Comment:
Indeed, one might phrase this as:

```
(defn- put-and-close! [chan close? value]
  (>! chan value)
  (when close?
    (a/close! chan)))

(defn consume-as-closed [chan]
  (let [the-error (make-the-error)]
    (go-loop [[_ _ result close?] (<! chan)]
      (when result
        (put-and-close! result close? [nil the-error])
        (recur)))))

…
(loop []
  (when-let [token …]
    (if @(:closed? conn)
      ;; Either we closed correctly, and we should make sure our input
      ;; channel is empty, or someone screwed up.
      (if (closed? (:transact-chan conn))
        (consume-as-closed (:transact-chan conn))
        (raise "DB closed under our transact chan!"))

      ;; Otherwise, consume one item, return the token, and loop.
      (do
        (let [pair …]
          …
          ;; Don't bother checking for chan closed.
          (put-and-close! result close? pair))
        (>! token-chan token)
        (recur)))))
```